File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.py
from collections import deque
def solution(x, y, n):
    if x == y: return 0
    visit = {x}
    q = deque([[x+n,1],[x*2,1],[x*3,1]])
    while q:
        val,cnt = q.popleft()
        if val > y: continue
        elif val == y: return cnt
        for i in [val+n,val*2,val*3]:
            if y >= i and i not in visit:
                visit.add(i) 
                q.append([i,cnt+1])
    return -1

File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.java
import java.util.*;

class Solution {
    public int solution(int x, int y, int n) {
        Queue<int[]> li = new LinkedList<>() {{
            add(new int[]{x,0});
        }};
        Set<Integer> visit = new HashSet<>() {{
            add(x);
        }};
        while (!li.isEmpty()) {
            int[] now = li.poll();
            if (now[0] == y) return now[1];
            for (int num: new int[]{now[0]+n,now[0]*2,now[0]*3}) {
                if (num<=y && !visit.contains(num)) {
                    visit.add(num);
                    li.add(new int[]{num,now[1]+1});
                }
            }
        }
        return -1;
    }
}